### PR TITLE
feat: Update IaC Interface to Match New Design

### PR DESCRIPTION
**Pull request by mildaniel**
### Change List:
- Renamed `class Project` to `class SamCliProject` to match design doc.
- Added `class ProjectFindRule` and `class Context`
- `class IaCPlugin` -> `class IaCPluginInterface` and made changes according to doc
- Added `class IaCPluginDefinition`

Changes start on line 796

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by wchengru — APPROVED** [review]

**Comment by wchengru** [comment]
Please resolve the pylint checks to unblock `make pr`.

File: samcli/lib/iac/plugins_interfaces.py [hunk 1]
```diff
@@ -0,0 +1,902 @@
...
+    @abc.abstractmethod
+    def project_detected(self, path: str) -> bool:
+        """
+        use the defined project_detector function to detect the type of a project
+        return true if a valid project is detected
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def valid_context(self, options: List[Any]) -> bool:
```
Comment (mildaniel) [inline samcli/lib/iac/plugins_interfaces.py:888]: Is `List[Any]` the correct type for click options?

Comment (wchengru) [reply samcli/lib/iac/plugins_interfaces.py:888]: It's supposed to be map of option-value pair, but leave for extra valiations.

Comment (mildaniel) [reply samcli/lib/iac/plugins_interfaces.py:888]: `Dict[Any, Any]`?

Comment (wchengru) [reply samcli/lib/iac/plugins_interfaces.py:888]: Agree with that.

**Review by wchengru — APPROVED** [review]

**Review by hawflau — APPROVED** [review]
LGTM!

**Pull request merged**